chain/network/src/peer_manager/testonly.rs

```diff
@@ -166,6 +168,47 @@ impl ActorHandler {
         }
     }
 
+    pub async fn force_connect_to(&self, peer_info: &PeerInfo, tier: tcp::Tier) {
+        let addr = self.actix.addr.clone();
+        let events = self.events.clone();
+        let peer_info = peer_info.clone();
+
+        loop {
+            let stream = tcp::Stream::connect(&peer_info, tier).await.unwrap();
+            let mut events = events.from_now();
+            let stream_id = stream.id();
+            addr.do_send(PeerManagerMessageRequest::OutboundTcpConnect(stream).with_span_context());
+            let success = events
+                .recv_until(|ev| match ev {
+                    Event::PeerManager(PME::HandshakeCompleted(ev))
+                        if ev.stream_id == stream_id =>
+                    {
+                        Some(true)
+                    }
+                    Event::PeerManager(PME::ConnectionClosed(ev)) if ev.stream_id == stream_id => {
+                        Some(
+                            ev.reason
+                                == ClosingReason::RejectedByPeerManager(
+                                    RegisterPeerError::PoolError(
+                                        connection::PoolError::AlreadyConnected,
+                                    ),
+                                )
+                                || ev.reason
+                                    == ClosingReason::OutboundNotAllowed(
+                                        connection::PoolError::AlreadyConnected,
+                                    ),
+                        )
+                    }
+                    _ => None,
```

Comment:
Why do you need a loop here? Do you expect some connection to fail at first try?
Also you do not set config.allow_outbound, so no spontaneous connections should be established, right?
Why do you expect AlreadyConnected then?

Reply:
I encountered some issues which I now understand to be timing-related. The old integration test handles them by waiting 50 ms after making each connection ([link](https://github.com/near/nearcore/pull/8168/files#diff-b1d740a0febe6d963975f850db2bf729ecf716cb2861f5e803e13f5bda7ff565L47)). As an example, if node 0 has 3 connections and does not finish dropping down to 2 connections before a 4th node attempts to connect, it will reject the new connection due to the configured `max_num_peers = 3`.

To resolve it I've removed the loop and instead inserted some `pm0.wait_for_num_connected_peers(2).await;` in the appropriate places.

However, there are still some non-deterministic occurrences of AlreadyConnected which I am trying to track down.

Reply:
A system-seeded rng is used to pick the connections to drop ([link](https://github.com/near/nearcore/blob/78ec412a846e8bb8182bcd58f08c003504ea1877/chain/network/src/peer_manager/peer_manager_actor.rs#L469)). The way this test is written the connections from nodes 2,3,4 can by chance survive until they are re-attempted.

I propose rewriting the test a bit; we can check pm0's connected peers and always select a non-connected peer to attempt the next connection. It still achieves the goal of testing that connections to archival nodes are retained preferentially (this test's intent was not to check that peer manager rejects duplicate connections).

Reply:
One more tweak; depending on timing of `monitor_peers_trigger`, peer manager might drop an inbound connection  "gracefully" even before the handshake is completed ([link](https://github.com/near/nearcore/blob/78ec412a846e8bb8182bcd58f08c003504ea1877/chain/network/src/peer_manager/peer_manager_actor.rs#L548)). However, the existing `connect_to` in peer_manager/testonly panics if it does not see a HandshakeCompleted event. I've switched this test to use a new `send_outbound_connect` so that the other peer managers simply initiate connections without waiting for HandshakeCompleted. We still check for a closed connection on `pm0` side for `ClosingReason::PeerManager` which should maintain correctness of this test.